blogpost: Pod Failure Policy for Jobs Goes GA (#46896)
kubernetes/website: content/en/blog/_posts/2024-08-19-pod-failure-policy-ga.md

this field would often lead to increased operating costs when Pods failed
because of non-retriable problems like bugs in code or database connectivity
issues, because a large backoffLimit value would require waiting for large
numbers of Pod failures before the Job failed.

Pod Failure Policy extends the failure mechanism to help you to reduce the costs
of running computational workloads in the following ways:

- Gives you control over when a Job fails. For example, you can fail the Job as
  soon as a non-retriable Pod failure occurs.
- Allowing you to run on more affordable machines without increasing the
  `backoffLimit` field by ignoring Pod failures caused by specific events. For
  example, you can ignore Pod failures that occur because of infrastructure
  failures like [graceful node shutdown](/docs/concepts/cluster-administration/node-shutdown/#graceful-node-shutdown).

## How it works

This feature introduces a condition, named `DisruptionTarget`, that Kubernetes
adds to a Pod when it fails because of a disruption. The condition contains one
of the following reasons for the Pod failure:

- `PreemptionByKubeScheduler` - Pod preempted by kube-scheduler

> - `DeletionByTaintManager` - Pod evicted by kube-controller-manager due to taints

Author: Yes, this is rephrased now, including the NoExecute mention

- `EvictionByEvictionAPI` - Pod deleted by Eviction API
- `DeletionByPodGC` - an orphaned Pod deleted by pod GC
- `TerminationByKubelet` - Pod terminated due to graceful node shutdown or
  preemption for critical pods.

You specify a Pod failure policy in the Job specification. It is represented
as a list of rules, each of which consists of a match condition and a
corresponding action.

Kubernetes evaluates these rules in order, matching them against end states of
failed Pods, based on the match conditions that you specify, which can look for
one of the following properties:
- Container exit codes: the `onExitCodes` property.
- Pod conditions: the `onPodConditions` property. These conditions can look for
  any of the [built-in Pod condition types](/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions)
  or for a custom Pod condition that you add using mechanisms like webhooks or external controllers.

For each rule, you can specify one of the following actions to take when a Pod
matches the specified match condition:

> - `Ignore` - don't count the failure towards the `.backoffLimit`

Author: correct, updated.

## Related work

We are pleased to observe the follow up work based on the concepts introduced
in this work, comprising:
- [Add more granular failure reason for Job PodFailurePolicy](https://github.com/kubernetes/enhancements/issues/4443)
- JobSet [Feature: Configurable Failure Policy API](https://github.com/kubernetes-sigs/jobset/issues/262)
- Proposal [Disrupted Pods should be removed from endpoints](https://docs.google.com/document/d/1t25jgO_-LRHhjRXf4KJ5xY_t8BZYdapv7MDAxVGY6R8)
- Support for Pod Failure Policy via JobSet in [Kubeflow Training v2](https://github.com/kubeflow/training-operator/pull/2171)

## Learn more

- Read the user-facing documentation for
  [Pod failure policy](/docs/concepts/workloads/controllers/job/#pod-failure-policy) and
  [Backoff limit per index](/docs/concepts/workloads/controllers/job/#backoff-limit-per-index)
- Read the tasks page on
  [Handling retriable and non-retriable pod failures with Pod failure policy](https://kubernetes.io/docs/tasks/job/pod-failure-policy/)
- Read the KEP for [Pod failure policy](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures)

> 

Reviewer: Additionally, it might be better to mention the Pod Disruption Conditions documentation: https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#pod-disruption-conditions

WDYT?

Author: Good idea, added

---
layout: blog
title: "Kubernetes 1.31: Pod failure policy goes GA"
date: 2024-08-19
slug: kubernetes-1-31-pod-failure-policy-stable
author: >
  Michał Woźniak (Google)
  Shannon Kularathna (Google)
---

> This post describes Pod Failure Policy, which graduates to stable in Kubernetes

Reviewer 2: (nit)
```suggestion
This post describes _pod failure policy_, which graduates to stable in Kubernetes
```

1.31, and how to use it in your Jobs.

> ## About Pod Failure Policy

Reviewer 2: (nit)
```suggestion
## About pod failure policy
```

> When running a workload on Kubernetes, things can go wrong resulting in Pod
> failures. Still, you expect your workload to tolerate transient failures and
> continue towards successful completion.
> 
> To compensate for transient issues, Kubernetes Jobs include the `backoffLimit`
> field, allowing you to specify the number of Pod failures tolerated during
> workload execution.
> 
> However, relying solely on this field can lead to unnecessary operating costs,
> as setting a large `backoffLimit` value results in excessive Pod restarts.
> 
> This becomes particularly problematic when running large-scale Jobs with
> thousands of long-running Pods across thousands of Nodes.

Reviewer 3: ```suggestion
Pod failures might occur when you run workloads on Kubernetes. However,
your workloads should ideally tolerate transient Pod failures and continue
towards successful completion.

To allow for these transient failures, Kubernetes Jobs include the `backoffLimit`
field, which lets you specify a number of Pod failures that you're willing to tolerate
during Job execution. However, if you set a large value for the `backoffLimit` field
and rely solely on this field, you might notice unnecessary increases in operating
costs as Pods restart excessively until the backoffLimit is met.

This becomes particularly problematic when running large-scale Jobs with
thousands of long-running Pods across thousands of nodes.
```

> The Pod Failure Policy extends the backoff limit mechanism to help you reduce

Reviewer 2: ```suggestion
Through pod failure policy, Kubernetes uses an extended backoff limit mechanism; this tries to help you reduce
```
?

Author: I'm not convinced this reads better, wdyt @shannonxtreme ?

Author: For now I have aligned the casing as "Pod failure policy" since this is the most consistent version with the rest of the blog.

Reviewer 3: Hm. Maybe 

"Through Pod failure policy, you get more control over the Kubernetes backoffLimit mechanism, which helps you to reduce..."? 

Is the issue with the current sentence the word "extends"?

Author: I think I like the current version for conciseness, but out of the two proposals, I prefer the latter. wdyt @sftim ?

costs in the following ways:

- Gives you control to fail the Job as soon as a non-retriable Pod failure occurs.
- Allows you to ignore retriable errors without increasing the `backoffLimit` field.

> For example, it allows you to run your workload on more affordable spot machines
> by ignoring Pod failures caused by
> [graceful node shutdown](/docs/concepts/cluster-administration/node-shutdown/#graceful-node-shutdown).

Reviewer 3: ```suggestion
For example, you can use a Pod failure policy to run your workload on more affordable spot machines
by ignoring Pod failures caused by
[graceful node shutdown](/docs/concepts/cluster-administration/node-shutdown/#graceful-node-shutdown).
```

> The policy allows you to distinguish between retriable and non-retriable Pod
> failures based on container exit codes and Pod conditions of a failed Pod.

Reviewer 3: ```suggestion
The policy allows you to distinguish between retriable and non-retriable Pod
failures based on container exit codes or Pod conditions in a failed Pod.
```

## How it works

You specify a Pod failure policy in the Job specification, represented as a list
of rules.

> For each rule you define match requirements based on either:
> - Container exit codes using the `onExitCodes` property.
> - Pod conditions using the `onPodConditions` property.

Reviewer 3: ```suggestion
For each rule you define _match requirements_ based on one of the following properties:

- Container exit codes: the `onExitCodes` property.
- Pod conditions: the `onPodConditions` property.
```

Additionally, for each rule, you specify one of the following actions to take
when a Pod matches the rule:

> - `Ignore` - Do not count the failure towards the `backoffLimit` or `backoffLimitPerIndex`.
> - `FailJob` - Fail the entire Job and terminate all running Pods.
> - `FailIndex` - Fail the index corresponding to the failed Pod
>   [Backoff limit per index](/docs/concepts/workloads/controllers/job/#backoff-limit-per-index),
> - `Count` - Count the failure towards the `backoffLimit` or `backoffLimitPerIndex`.
>   This is the default behavior.

Reviewer 3: ```suggestion
- `Ignore`: Do not count the failure towards the `backoffLimit` or `backoffLimitPerIndex`.
- `FailJob`: Fail the entire Job and terminate all running Pods.
- `FailIndex`: Fail the index corresponding to the failed Pod
  [Backoff limit per index](/docs/concepts/workloads/controllers/job/#backoff-limit-per-index),
- `Count`: Count the failure towards the `backoffLimit` or `backoffLimitPerIndex`.
  This is the default behavior.
```

  This is the default behavior.

> When Pod failures occur in a running Job, Kubernetes matches the

Reviewer 4: I know that this blog post is about policy, but it may worth mentioning that Pod restartPolicy must be Never. To highlight that the node is not making the restart decision.

Author: Thanks for the suggestion. This is indeed an important design limitation. It is mentioned on the [Pod Failure Policy page](https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-failure-policy): "if you want to use a .spec.podFailurePolicy field for a Job, you must also define that Job's pod template with .spec.restartPolicy set to Never.".

We could consider saying something like: 
``` 
When you specify the Pod failure policy you must also define the Job's Pod template with
`restartPolicy: Never`. This is required to ensure there are no race conditions between
Kubelet and Job controller when counting Pod failures" (at the bottom of the "How it works".
```

OTOH, I think this might be too detailed for the tone of the blog post. WDYT @atiratree @shannonxtreme ?

Reviewer 5: I think it could be useful to have, especially since we do not have a full example with the Job/Pod object in the blog.

> When you run your Job and Pod failures occur, then Kubernetes matches the
> failed Pod status against the list of Pod failure policy rules, in the specified
> order, and takes the corresponding actions for the first matched rule.

Reviewer 3: ```suggestion
When Pod failures occur in a running Job, Kubernetes matches the
failed Pod status against the list of Pod failure policy rules, in the specified
order, and takes the corresponding actions for the first matched rule.
```

> ### Pod disruption conditions

Reviewer 3: ```suggestion
### Kubernetes-initiated Pod disruptions
```

Something like this, that talks more about what the user is looking to match for

  [graceful node shutdown](/docs/concepts/cluster-administration/node-shutdown/#graceful-node-shutdown),
  [node pressure eviction](/docs/concepts/scheduling-eviction/node-pressure-eviction/)
  or preemption for [system critical pods](/docs/tasks/administer-cluster/guaranteed-scheduling-critical-addon-pods/).

### Example

The Pod failure policy snippet below demonstrates an example use:

```yaml
podFailurePolicy:
  rules:
  - action: Ignore
    onPodConditions:
    - type: DisruptionTarget
  - action: FailJob
    onPodConditions:
    - type: ConfigIssue
  - action: FailJob
    onExitCodes:
      operator: In
      values: [ 42 ]
```

This example, the Pod failure policy does the following:

- Ignores any failed Pods that have the built-in `DisruptionTarget`
  condition. These Pods don't count towards Job backoff limits.
- Fails the Job if any failed Pods have the custom user-supplied
  `ConfigIssue` condition, which was added either by a custom controller or webhook.
- Fails the Job if any containers exited with the reserved exit code 42
- Counts all other Pod failures towards the default `backoffLimit` (or
  `backoffLimitPerIndex` if used).

## Learn more

> - Read [Handling retriable and non-retriable pod failures with Pod failure policy](/docs/tasks/job/pod-failure-policy/)

Reviewer 3: Recommend flipping structure

"For a hands-on guide to using Pod failure policy, see <link>"

Author: done

>   for a hands-on experience of using Pod Failure Policy

Reviewer 3: ```suggestion
  for a hands-on guide to using Pod Failure Policy
```

failures based on container exit codes and Pod conditions of a failed Pod.

## How it works

You specify a Pod failure policy in the Job specification, represented as a list
of rules.

For each rule you define match requirements based on either:
- Container exit codes using the `onExitCodes` property.
- Pod conditions using the `onPodConditions` property.

Additionally, for each rule, you specify one of the following actions to take
when a Pod matches the rule:
- `Ignore` - Do not count the failure towards the `backoffLimit` or `backoffLimitPerIndex`.
- `FailJob` - Fail the entire Job and terminate all running Pods.
- `FailIndex` - Fail the index corresponding to the failed Pod
  [Backoff limit per index](/docs/concepts/workloads/controllers/job/#backoff-limit-per-index),
- `Count` - Count the failure towards the `backoffLimit` or `backoffLimitPerIndex`.
  This is the default behavior.

When you run your Job and Pod failures occur, then Kubernetes matches the
failed Pod status against the list of Pod failure policy rules, in the specified
order, and takes the corresponding actions for the first matched rule.

### Pod disruption conditions

> In order to allow matching Pod Failure Policy against failures due to
> disruptions initiated by Kubernetes, this feature comes with a new Pod
> condition, called `DisruptionTarget`.
> 
> This condition is added by Kubernetes it case of retriable
> [disruption scenarios](/docs/concepts/workloads/pods/disruptions/#pod-disruption-conditions).
> The condition contains one of the following reasons corresponding to the
> scenarios:

Reviewer 3: ```suggestion
To allow matching Pod failure policy rules against failures caused by
disruptions initiated by Kubernetes, this feature introduces the `DisruptionTarget`
Pod condition.

This condition is added by Kubernetes when a Pod fails because of a retriable
[disruption scenario](/docs/concepts/workloads/pods/disruptions/#pod-disruption-conditions).
The `DisruptionTarget` condition contains one of the following reasons that corresponds to
these disruption scenarios:
```

> - `PreemptionByKubeScheduler` - [Preemption](/docs/concepts/scheduling-eviction/pod-priority-preemption)
>    by `kube-scheduler` in order to accommodate a new Pod with a higher priority.
> - `DeletionByTaintManager` - Pod is due to be deleted by
>    `kube-controller-manager` due to `NoExecute` [taint](/docs/concepts/scheduling-eviction/taint-and-toleration/)
>    that the Pod does not tolerate.
> - `EvictionByEvictionAPI` - Pod has been marked for
>    [eviction using the Kubernetes API](/docs/concepts/scheduling-eviction/api-eviction/).
> - `DeletionByPodGC` - Pod, that is bound to a no longer existing Node, is due to
>    be deleted by [Pod garbage collection](/docs/concepts/workloads/pods/pod-lifecycle/#pod-garbage-collection).
> - `TerminationByKubelet` - Pod terminated due to
>   [graceful node shutdown](/docs/concepts/cluster-administration/node-shutdown/#graceful-node-shutdown),
>   [node pressure eviction](/docs/concepts/scheduling-eviction/node-pressure-eviction/)
>   or preemption for [system critical pods](/docs/tasks/administer-cluster/guaranteed-scheduling-critical-addon-pods/).

Reviewer 3: ```suggestion
- `PreemptionByKubeScheduler`: [Preemption](/docs/concepts/scheduling-eviction/pod-priority-preemption)
   by `kube-scheduler` to accommodate a new Pod that has a higher priority.
- `DeletionByTaintManager` - the Pod is due to be deleted by
   `kube-controller-manager` due to a `NoExecute` [taint](/docs/concepts/scheduling-eviction/taint-and-toleration/)
   that the Pod doesn't tolerate.
- `EvictionByEvictionAPI` - the Pod has been marked for
   [eviction using the Kubernetes API](/docs/concepts/scheduling-eviction/api-eviction/).
- `DeletionByPodGC` - the Pod is bound to a node that no longer exists, and is due to
   be deleted by [Pod garbage collection](/docs/concepts/workloads/pods/pod-lifecycle/#pod-garbage-collection).
- `TerminationByKubelet` - the Pod was terminated by
  [graceful node shutdown](/docs/concepts/cluster-administration/node-shutdown/#graceful-node-shutdown),
  [node pressure eviction](/docs/concepts/scheduling-eviction/node-pressure-eviction/)
  or preemption for [system critical pods](/docs/tasks/administer-cluster/guaranteed-scheduling-critical-addon-pods/).
```

### Example

The Pod failure policy snippet below demonstrates an example use:

```yaml
podFailurePolicy:
  rules:
  - action: Ignore
    onPodConditions:
    - type: DisruptionTarget
  - action: FailJob
    onPodConditions:
    - type: ConfigIssue
  - action: FailJob
    onExitCodes:
      operator: In
      values: [ 42 ]
```

This example, the Pod failure policy does the following:

> - ignores any failed Pods that have the built-in `DisruptionTarget`
>   condition. These Pods don't count towards Job backoff limits.
> - fails the Job if any failed Pods have the custom user-supplied
>   `ConfigIssue` condition, which was added either by a custom controller or webhook.
> - fails the Job if any containers exited with the reserved exit code 42
> - counts all other Pod failures towards the default `backoffLimit` (or
>   `backoffLimitPerIndex` if used).

Reviewer 3: ```suggestion
- Ignores any failed Pods that have the built-in `DisruptionTarget`
  condition. These Pods don't count towards Job backoff limits.
- Fails the Job if any failed Pods have the custom user-supplied
  `ConfigIssue` condition, which was added either by a custom controller or webhook.
- Fails the Job if any containers exited with the reserved exit code 42
- Counts all other Pod failures towards the default `backoffLimit` (or
  `backoffLimitPerIndex` if used).
```

## Learn more

- Read the documentation for
  [Pod failure policy](/docs/concepts/workloads/controllers/job/#pod-failure-policy) and
  [Backoff limit per index](/docs/concepts/workloads/controllers/job/#backoff-limit-per-index)
- Read the documentation for
  [Pod disruption conditions](/docs/concepts/workloads/pods/disruptions/#pod-disruption-conditions)
- Read the tasks page on
  [Handling retriable and non-retriable pod failures with Pod failure policy](/docs/tasks/job/pod-failure-policy/)
- Read the KEP for [Pod failure policy](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures)

## Related work

> We are pleased to observe the follow up work based on the concepts introduced
> in this work, comprising:

Reviewer 3: ```suggestion
Based on the concepts introduced by Pod failure policy, the following additional work is in progress:
```
Something that's not "we're pleased to", basically

these disruption scenarios:

- `PreemptionByKubeScheduler`: [Preemption](/docs/concepts/scheduling-eviction/pod-priority-preemption)
   by `kube-scheduler` to accommodate a new Pod that has a higher priority.
- `DeletionByTaintManager` - the Pod is due to be deleted by
   `kube-controller-manager` due to a `NoExecute` [taint](/docs/concepts/scheduling-eviction/taint-and-toleration/)
   that the Pod doesn't tolerate.
- `EvictionByEvictionAPI` - the Pod has been marked for
   [eviction using the Kubernetes API](/docs/concepts/scheduling-eviction/api-eviction/).
- `DeletionByPodGC` - the Pod is bound to a node that no longer exists, and is due to
   be deleted by [Pod garbage collection](/docs/concepts/workloads/pods/pod-lifecycle/#pod-garbage-collection).
- `TerminationByKubelet` - the Pod was terminated by
  [graceful node shutdown](/docs/concepts/cluster-administration/node-shutdown/#graceful-node-shutdown),
  [node pressure eviction](/docs/concepts/scheduling-eviction/node-pressure-eviction/)
  or preemption for [system critical pods](/docs/tasks/administer-cluster/guaranteed-scheduling-critical-addon-pods/).

> ### Examples

Author: I tentaively split the single example into 3 smaller in this commit: https://github.com/kubernetes/website/pull/46896/commits/227acb546c77615fac490ba5718f0bb6a658edbf. 

Motivation:
- to also describe FailIndex for backoff limit per index
- to make smaller, more focused snippets, which are easier to describe

WDYT @tenzen-y @atiratree @shannonxtreme ?

Reviewer 3: I say no. The task page has that already, split into three, only thing that's missing is the index one. Blogs go out of date fast, docs don't, so recommend that you save details like this for the docs and rely on them. If content from your commit isn't already in https://kubernetes.io/docs/tasks/job/pod-failure-policy/ put it there at some point

Author: Given the feedback I reverted the last commit, but I'm happy to reconsider some parts of it if other find it useful - the link to the commit still works.

Reviewer: TBH, I don't have a strong view of this example, but @shannonxtreme opinion sounds reasonable.
In general, blog articles are easy to outdated.

Reviewer: /lgtm

Author: sgtm, I think we add extending the tutorial for backoff limit per index + pod failure policy, as a graduation criteria for backoff limit per index to stable. WDYT?

Reviewer: Does it mean we want to add some more examples for the back off limit per index + pod failure policy to the documentation, not blog?

It sounds attractive, but maybe we can revisit this discussion when the backoff limit per index feature graduate to stable.

Author: > Does it mean we want to add some more examples for the back off limit per index + pod failure policy to the documentation, not blog?

Yeah, this is what I was considering

> It sounds attractive, but maybe we can revisit this discussion when the backoff limit per index feature graduate to stable.

sgtm

Reviewer: Sound good to me.
